src/Features/Core/Portable/CodeRefactorings/MoveType/AbstractMoveTypeService.Editor.cs

```diff
@@ -0,0 +1,272 @@
+                    return RenameFileToMatchTypeName(solution);
+                }
+                else if (_renameType)
+                {
+                    return await RenameTypeToMatchFileAsync(solution).ConfigureAwait(false);
+                }
+
+                var documentName = _fromDialog
+                    ? _moveTypeOptions.NewFileName
+                    : _state.TargetFileNameCandidate + _state.TargetFileExtension;
+
+                return await MoveTypeToNewFileAsync(documentName).ConfigureAwait(false);
+            }
+
+            private async Task<IEnumerable<CodeActionOperation>> MoveTypeToNewFileAsync(string documentName)
+            {
+                // fork source document, keep required type/namespace hierarchy and add it to a new document
+                var projectToBeUpdated = SemanticDocument.Document.Project;
+                var newDocumentId = DocumentId.CreateNewId(projectToBeUpdated.Id, documentName);
+
+                var solutionWithNewDocument = await AddNewDocumentWithTypeDeclarationAsync(
+                    SemanticDocument, documentName, newDocumentId, _state.TypeNode, _cancellationToken).ConfigureAwait(false);
+
+                // Get the original source document again, from the latest forked solution.
+                var sourceDocument = solutionWithNewDocument.GetDocument(SemanticDocument.Document.Id);
+
+                // update source document to add partial modifiers to type chain
+                // and/or remove type declaration from original source document.
+                var solutionWithBothDocumentsUpdated = await UpdateSourceDocumentAsync(
+                      sourceDocument, _state.TypeNode, _cancellationToken).ConfigureAwait(false);
+
+                return new CodeActionOperation[] { new ApplyChangesOperation(solutionWithBothDocumentsUpdated) };
+            }
+
+            private async Task<IEnumerable<CodeActionOperation>> RenameTypeToMatchFileAsync(Solution solution)
+            {
+                var symbol = _state.SemanticDocument.SemanticModel.GetDeclaredSymbol(_state.TypeNode, _cancellationToken) as INamedTypeSymbol;
+                var newSolution = await Renamer.RenameSymbolAsync(solution, symbol, _state.DocumentName, SemanticDocument.Document.Options, _cancellationToken).ConfigureAwait(false);
+                return new CodeActionOperation[] { new ApplyChangesOperation(newSolution) };
+            }
+
+            private IEnumerable<CodeActionOperation> RenameFileToMatchTypeName(Solution solution)
+            {
+                var text = SemanticDocument.Text;
+                var oldDocumentId = SemanticDocument.Document.Id;
+                var newDocumentId = DocumentId.CreateNewId(SemanticDocument.Document.Project.Id, _state.TargetFileNameCandidate);
+
+                var newSolution = solution.RemoveDocument(oldDocumentId);
+                newSolution = newSolution.AddDocument(newDocumentId, _state.TargetFileNameCandidate, text);
```

Comment:
I'm not really happy with this approach to accomplish a file rename.  @mattwar @jasonmalinowski Could we instead just expose a RenameDocument primitive on the workspace/solution itself to better capture the intent of what's happening here?